{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " # Loading an HCA matrix into Bioconductor\n",
    " \n",
    "This vignette illustrates requesting an expression matrix from the HCA matrix service and loading it into a Bioconductor R object.\n",
    "\n",
    "First, install and import some dependencies:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/nl/dgln31tj7l35g879d6f_tjtc0000gn/T//RtmpKQd50E/downloaded_packages\n",
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/nl/dgln31tj7l35g879d6f_tjtc0000gn/T//RtmpKQd50E/downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bioconductor version 3.9 (BiocManager 1.30.4), R 3.6.0 (2019-04-26)\n",
      "Installing package(s) 'LoomExperiment'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/nl/dgln31tj7l35g879d6f_tjtc0000gn/T//RtmpKQd50E/downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Update old packages: 'dplyr', 'googleAuthR', 'pillar'\n",
      "Loading required package: SingleCellExperiment\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: DelayedArray\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "Loading required package: rhdf5\n",
      "Loading required package: rtracklayer\n",
      "\n",
      "Attaching package: ‘httr’\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    content\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"downloader\")\n",
    "install.packages(\"BiocManager\")\n",
    "\n",
    "library(\"downloader\")\n",
    "library(\"BiocManager\")\n",
    "\n",
    "BiocManager::install(\"LoomExperiment\")\n",
    "library(LoomExperiment)\n",
    "\n",
    "library(\"httr\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we're going to make some requests to describe what fields and values we can filter on when we're selecting our matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li>'cell_suspension.provenance.document_id'</li>\n",
       "\t<li>'genes_detected'</li>\n",
       "\t<li>'specimen_from_organism.provenance.document_id'</li>\n",
       "\t<li>'specimen_from_organism.genus_species.ontology'</li>\n",
       "\t<li>'specimen_from_organism.genus_species.ontology_label'</li>\n",
       "\t<li>'donor_organism.human_specific.ethnicity.ontology'</li>\n",
       "\t<li>'donor_organism.human_specific.ethnicity.ontology_label'</li>\n",
       "\t<li>'donor_organism.diseases.ontology'</li>\n",
       "\t<li>'donor_organism.diseases.ontology_label'</li>\n",
       "\t<li>'donor_organism.development_stage.ontology'</li>\n",
       "\t<li>'donor_organism.development_stage.ontology_label'</li>\n",
       "\t<li>'derived_organ_ontology'</li>\n",
       "\t<li>'derived_organ_label'</li>\n",
       "\t<li>'derived_organ_part_ontology'</li>\n",
       "\t<li>'derived_organ_part_label'</li>\n",
       "\t<li>'library_preparation_protocol.provenance.document_id'</li>\n",
       "\t<li>'library_preparation_protocol.input_nucleic_acid_molecule.ontology'</li>\n",
       "\t<li>'library_preparation_protocol.input_nucleic_acid_molecule.ontology_label'</li>\n",
       "\t<li>'library_preparation_protocol.library_construction_method.ontology'</li>\n",
       "\t<li>'library_preparation_protocol.library_construction_method.ontology_label'</li>\n",
       "\t<li>'library_preparation_protocol.end_bias'</li>\n",
       "\t<li>'library_preparation_protocol.strand'</li>\n",
       "\t<li>'project.provenance.document_id'</li>\n",
       "\t<li>'project.project_core.project_short_name'</li>\n",
       "\t<li>'project.project_core.project_title'</li>\n",
       "\t<li>'analysis_protocol.provenance.document_id'</li>\n",
       "\t<li>'dss_bundle_fqid'</li>\n",
       "\t<li>'analysis_protocol.protocol_core.protocol_id'</li>\n",
       "\t<li>'analysis_working_group_approval_status'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item 'cell\\_suspension.provenance.document\\_id'\n",
       "\\item 'genes\\_detected'\n",
       "\\item 'specimen\\_from\\_organism.provenance.document\\_id'\n",
       "\\item 'specimen\\_from\\_organism.genus\\_species.ontology'\n",
       "\\item 'specimen\\_from\\_organism.genus\\_species.ontology\\_label'\n",
       "\\item 'donor\\_organism.human\\_specific.ethnicity.ontology'\n",
       "\\item 'donor\\_organism.human\\_specific.ethnicity.ontology\\_label'\n",
       "\\item 'donor\\_organism.diseases.ontology'\n",
       "\\item 'donor\\_organism.diseases.ontology\\_label'\n",
       "\\item 'donor\\_organism.development\\_stage.ontology'\n",
       "\\item 'donor\\_organism.development\\_stage.ontology\\_label'\n",
       "\\item 'derived\\_organ\\_ontology'\n",
       "\\item 'derived\\_organ\\_label'\n",
       "\\item 'derived\\_organ\\_part\\_ontology'\n",
       "\\item 'derived\\_organ\\_part\\_label'\n",
       "\\item 'library\\_preparation\\_protocol.provenance.document\\_id'\n",
       "\\item 'library\\_preparation\\_protocol.input\\_nucleic\\_acid\\_molecule.ontology'\n",
       "\\item 'library\\_preparation\\_protocol.input\\_nucleic\\_acid\\_molecule.ontology\\_label'\n",
       "\\item 'library\\_preparation\\_protocol.library\\_construction\\_method.ontology'\n",
       "\\item 'library\\_preparation\\_protocol.library\\_construction\\_method.ontology\\_label'\n",
       "\\item 'library\\_preparation\\_protocol.end\\_bias'\n",
       "\\item 'library\\_preparation\\_protocol.strand'\n",
       "\\item 'project.provenance.document\\_id'\n",
       "\\item 'project.project\\_core.project\\_short\\_name'\n",
       "\\item 'project.project\\_core.project\\_title'\n",
       "\\item 'analysis\\_protocol.provenance.document\\_id'\n",
       "\\item 'dss\\_bundle\\_fqid'\n",
       "\\item 'analysis\\_protocol.protocol\\_core.protocol\\_id'\n",
       "\\item 'analysis\\_working\\_group\\_approval\\_status'\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. 'cell_suspension.provenance.document_id'\n",
       "2. 'genes_detected'\n",
       "3. 'specimen_from_organism.provenance.document_id'\n",
       "4. 'specimen_from_organism.genus_species.ontology'\n",
       "5. 'specimen_from_organism.genus_species.ontology_label'\n",
       "6. 'donor_organism.human_specific.ethnicity.ontology'\n",
       "7. 'donor_organism.human_specific.ethnicity.ontology_label'\n",
       "8. 'donor_organism.diseases.ontology'\n",
       "9. 'donor_organism.diseases.ontology_label'\n",
       "10. 'donor_organism.development_stage.ontology'\n",
       "11. 'donor_organism.development_stage.ontology_label'\n",
       "12. 'derived_organ_ontology'\n",
       "13. 'derived_organ_label'\n",
       "14. 'derived_organ_part_ontology'\n",
       "15. 'derived_organ_part_label'\n",
       "16. 'library_preparation_protocol.provenance.document_id'\n",
       "17. 'library_preparation_protocol.input_nucleic_acid_molecule.ontology'\n",
       "18. 'library_preparation_protocol.input_nucleic_acid_molecule.ontology_label'\n",
       "19. 'library_preparation_protocol.library_construction_method.ontology'\n",
       "20. 'library_preparation_protocol.library_construction_method.ontology_label'\n",
       "21. 'library_preparation_protocol.end_bias'\n",
       "22. 'library_preparation_protocol.strand'\n",
       "23. 'project.provenance.document_id'\n",
       "24. 'project.project_core.project_short_name'\n",
       "25. 'project.project_core.project_title'\n",
       "26. 'analysis_protocol.provenance.document_id'\n",
       "27. 'dss_bundle_fqid'\n",
       "28. 'analysis_protocol.protocol_core.protocol_id'\n",
       "29. 'analysis_working_group_approval_status'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "[1] \"cell_suspension.provenance.document_id\"\n",
       "\n",
       "[[2]]\n",
       "[1] \"genes_detected\"\n",
       "\n",
       "[[3]]\n",
       "[1] \"specimen_from_organism.provenance.document_id\"\n",
       "\n",
       "[[4]]\n",
       "[1] \"specimen_from_organism.genus_species.ontology\"\n",
       "\n",
       "[[5]]\n",
       "[1] \"specimen_from_organism.genus_species.ontology_label\"\n",
       "\n",
       "[[6]]\n",
       "[1] \"donor_organism.human_specific.ethnicity.ontology\"\n",
       "\n",
       "[[7]]\n",
       "[1] \"donor_organism.human_specific.ethnicity.ontology_label\"\n",
       "\n",
       "[[8]]\n",
       "[1] \"donor_organism.diseases.ontology\"\n",
       "\n",
       "[[9]]\n",
       "[1] \"donor_organism.diseases.ontology_label\"\n",
       "\n",
       "[[10]]\n",
       "[1] \"donor_organism.development_stage.ontology\"\n",
       "\n",
       "[[11]]\n",
       "[1] \"donor_organism.development_stage.ontology_label\"\n",
       "\n",
       "[[12]]\n",
       "[1] \"derived_organ_ontology\"\n",
       "\n",
       "[[13]]\n",
       "[1] \"derived_organ_label\"\n",
       "\n",
       "[[14]]\n",
       "[1] \"derived_organ_part_ontology\"\n",
       "\n",
       "[[15]]\n",
       "[1] \"derived_organ_part_label\"\n",
       "\n",
       "[[16]]\n",
       "[1] \"library_preparation_protocol.provenance.document_id\"\n",
       "\n",
       "[[17]]\n",
       "[1] \"library_preparation_protocol.input_nucleic_acid_molecule.ontology\"\n",
       "\n",
       "[[18]]\n",
       "[1] \"library_preparation_protocol.input_nucleic_acid_molecule.ontology_label\"\n",
       "\n",
       "[[19]]\n",
       "[1] \"library_preparation_protocol.library_construction_method.ontology\"\n",
       "\n",
       "[[20]]\n",
       "[1] \"library_preparation_protocol.library_construction_method.ontology_label\"\n",
       "\n",
       "[[21]]\n",
       "[1] \"library_preparation_protocol.end_bias\"\n",
       "\n",
       "[[22]]\n",
       "[1] \"library_preparation_protocol.strand\"\n",
       "\n",
       "[[23]]\n",
       "[1] \"project.provenance.document_id\"\n",
       "\n",
       "[[24]]\n",
       "[1] \"project.project_core.project_short_name\"\n",
       "\n",
       "[[25]]\n",
       "[1] \"project.project_core.project_title\"\n",
       "\n",
       "[[26]]\n",
       "[1] \"analysis_protocol.provenance.document_id\"\n",
       "\n",
       "[[27]]\n",
       "[1] \"dss_bundle_fqid\"\n",
       "\n",
       "[[28]]\n",
       "[1] \"analysis_protocol.protocol_core.protocol_id\"\n",
       "\n",
       "[[29]]\n",
       "[1] \"analysis_working_group_approval_status\"\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "r <- GET(\"https://matrix.data.humancellatlas.org/v1/filters\")\n",
    "content(r)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That's the list of metadata fields we can filter on when requesting the matrix. We can describe any of them with further API calls:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "$cell_counts\n",
      "$cell_counts$`Fetal/Maternal Interface`\n",
      "[1] 1\n",
      "\n",
      "$cell_counts$`Single cell RNAseq characterization of cell types produced over time in an in vitro model of human inhibitory interneuron differentiation.`\n",
      "[1] 1733\n",
      "\n",
      "$cell_counts$`Single cell transcriptome analysis of human pancreas`\n",
      "[1] 2544\n",
      "\n",
      "\n",
      "$field_description\n",
      "[1] \"A short name for the project.\"\n",
      "\n",
      "$field_name\n",
      "[1] \"project.project_core.project_short_name\"\n",
      "\n",
      "$field_type\n",
      "[1] \"categorical\"\n",
      "\n"
     ]
    }
   ],
   "source": [
    "r <- GET(\"https://matrix.data.humancellatlas.org/v1/filters/project.project_core.project_short_name\")\n",
    "print(content(r))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "$field_description\n",
      "[1] \"Count of genes with a non-zero count.\"\n",
      "\n",
      "$field_name\n",
      "[1] \"genes_detected\"\n",
      "\n",
      "$field_type\n",
      "[1] \"numeric\"\n",
      "\n",
      "$maximum\n",
      "[1] 13108\n",
      "\n",
      "$minimum\n",
      "[1] 358\n",
      "\n"
     ]
    }
   ],
   "source": [
    "r <- GET(\"https://matrix.data.humancellatlas.org/v1/filters/genes_detected\")\n",
    "print(content(r))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For categorical data, we see the number of cells associated with each category. For numeric, we see the range of value. If we want to request a matrix based on these metadata values, we can add them to the filter in the body of a POST request to the matrix service:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "$eta\n",
      "[1] \"\"\n",
      "\n",
      "$matrix_url\n",
      "[1] \"\"\n",
      "\n",
      "$message\n",
      "[1] \"Job started.\"\n",
      "\n",
      "$request_id\n",
      "[1] \"4748716f-772f-4716-8500-c8c21e4ad237\"\n",
      "\n",
      "$status\n",
      "[1] \"In Progress\"\n",
      "\n"
     ]
    }
   ],
   "source": [
    "payload = list(\n",
    "    filter =  list(\n",
    "          op = \"and\", \n",
    "          value = list(\n",
    "              list(op = \"=\", value = \"Single cell transcriptome analysis of human pancreas\",\n",
    "                   field = \"project.project_core.project_short_name\"),\n",
    "              list(op = \">=\", value = 300,\n",
    "                   field = \"genes_detected\")\n",
    "    )),\n",
    "    format = \"loom\"\n",
    ")\n",
    "r <- POST(\"https://matrix.data.humancellatlas.org/v1/matrix\", body = payload, encode = \"json\")\n",
    "response <- content(r)\n",
    "print(response)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That call responds right away and tells us that the matrix is being prepared. We can use the request_id to wait until the matrix is done."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "In Progress\n",
      "In Progress\n",
      "In Progress\n",
      "In Progress\n",
      "In Progress\n",
      "In Progress\n",
      "In Progress\n",
      "In Progress\n",
      "In Progress\n",
      "In Progress\n",
      "In Progress\n",
      "Complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "$eta\n",
      "[1] \"\"\n",
      "\n",
      "$matrix_url\n",
      "[1] \"https://s3.amazonaws.com/dcp-matrix-service-results-prod/4748716f-772f-4716-8500-c8c21e4ad237.loom\"\n",
      "\n",
      "$message\n",
      "[1] \"Request 4748716f-772f-4716-8500-c8c21e4ad237 has successfully completed. The resultant expression matrix is available for download at https://s3.amazonaws.com/dcp-matrix-service-results-prod/4748716f-772f-4716-8500-c8c21e4ad237.loom\"\n",
      "\n",
      "$request_id\n",
      "[1] \"4748716f-772f-4716-8500-c8c21e4ad237\"\n",
      "\n",
      "$status\n",
      "[1] \"Complete\"\n",
      "\n"
     ]
    }
   ],
   "source": [
    "request_id <- response[\"request_id\"]\n",
    "status <- response[\"status\"]\n",
    "message(status)\n",
    "while (status != \"Complete\") \n",
    "{\n",
    "    url = paste(\"https://matrix.data.humancellatlas.org/v1/matrix/\", request_id, sep = \"\")\n",
    "    r <- GET(url)\n",
    "    response <- content(r)\n",
    "    status = response[\"status\"]\n",
    "    message(status)\n",
    "    Sys.sleep(15)\n",
    "}\n",
    "print(response)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, that the matrix is ready, we can download it. The file we download is a loom-formatted matrix. Loom is the default matrix format, but others can be specified in the matrix request (csv, mtx)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix_file_url = unlist(response[\"matrix_url\"])\n",
    "\n",
    "download.file(url=matrix_file_url,\n",
    "              destfile='matrix.loom', method='curl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HCA Matrix Service Loom Output\n",
    "\n",
    "The Loom format is documented more fully, along with code samples, [here](https://linnarssonlab.org/loompy/index.html).\n",
    "\n",
    "Per Loom [conventions](https://linnarssonlab.org/loompy/conventions/index.html), columns in the loom-formatted expression matrix represent cells, and rows represent genes. The column and row attributes follow Loom conventions where applicable as well: `CellID` uniquely identifies a cell, `Gene` is a gene name, and `Accession` is an ensembl gene id.\n",
    "\n",
    "Descriptions of the remaining metadata fields are available at the [HCA Data Browser](https://prod.data.humancellatlas.org/metadata).\n",
    "\n",
    "And finally, we can `import` the loom file into a `Bioconductor::SingleCellLoomExperiment` object for further analysis in R."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: SingleCellLoomExperiment \n",
       "dim: 63925 2544 \n",
       "metadata(3): CreationDate LOOM_SPEC_VERSION last_modified\n",
       "assays(1): matrix\n",
       "rownames: NULL\n",
       "rowData names(7): Accession Gene ... featuretype isgene\n",
       "colnames: NULL\n",
       "colData names(30): CellID analysis_protocol.protocol_core.protocol_id\n",
       "  ... specimen_from_organism.genus_species.ontology_label\n",
       "  specimen_from_organism.provenance.document_id\n",
       "reducedDimNames(0):\n",
       "spikeNames(0):\n",
       "rowGraphs(0): NULL\n",
       "colGraphs(0): NULL"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scle <- import(\"./matrix.loom\", type=\"SingleCellLoomExperiment\")\n",
    "scle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `SingleCellLoomExperiment` also adheres to Loom [conventions](https://linnarssonlab.org/loompy/conventions/index.html) representing features as rows and samples as columns. Expression data is available via the `assays()` method, specifying a named assay."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<63925 x 2544> DelayedMatrix object of type \"double\":\n",
       "            [,1]    [,2]    [,3] ... [,2543] [,2544]\n",
       "    [1,]       0       0       0   .       0       0\n",
       "    [2,]       0       0       0   .       0       0\n",
       "    [3,]      11       0      37   .       0       6\n",
       "    [4,]       0       0       0   .     101       0\n",
       "    [5,]       0       0       0   .       0       0\n",
       "     ...       .       .       .   .       .       .\n",
       "[63921,]    0.00    0.00    0.00   .       0       2\n",
       "[63922,]    0.00    0.00    0.00   .       0       0\n",
       "[63923,]    0.00    0.00    0.00   .       0       0\n",
       "[63924,]    0.99    0.00    0.00   .       0       0\n",
       "[63925,]    0.00    0.00    0.00   .       0       0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "assays(scle)$matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Row and column attribute data are available through `rowData()` and `colData()` methods respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 63925 rows and 7 columns\n",
       "            Accession        Gene  chromosome featureend featurestart\n",
       "          <character> <character> <character>  <integer>    <integer>\n",
       "1     ENSG00000000003      TSPAN6        chrX  100639991    100627109\n",
       "2     ENSG00000000005        TNMD        chrX  100599885    100584802\n",
       "3     ENSG00000000419        DPM1       chr20   50958555     50934867\n",
       "4     ENSG00000000457       SCYL3        chr1  169894267    169849631\n",
       "5     ENSG00000000460    C1orf112        chr1  169854080    169662007\n",
       "...               ...         ...         ...        ...          ...\n",
       "63921 ENSG00000284744  AL591163.1        chr1    6770038      6767954\n",
       "63922 ENSG00000284745  AL589702.1        chr1    2968707      2960658\n",
       "63923 ENSG00000284746 AC068587.10        chr8   12601376     12601158\n",
       "63924 ENSG00000284747  AL034417.4        chr1    8005312      7991134\n",
       "63925 ENSG00000284748  AL513220.1        chr1   37607336     37596126\n",
       "                             featuretype    isgene\n",
       "                             <character> <integer>\n",
       "1                         protein_coding         1\n",
       "2                         protein_coding         1\n",
       "3                         protein_coding         1\n",
       "4                         protein_coding         1\n",
       "5                         protein_coding         1\n",
       "...                                  ...       ...\n",
       "63921                            lincRNA         1\n",
       "63922                            lincRNA         1\n",
       "63923 transcribed_unprocessed_pseudogene         1\n",
       "63924                      antisense_RNA         1\n",
       "63925      bidirectional_promoter_lncRNA         1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rowData(scle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DataFrame with 2544 rows and 30 columns\n",
       "                                   CellID\n",
       "                              <character>\n",
       "1    00ca0d37-b787-41a4-be59-2aff5b13b0bd\n",
       "2    0103aed0-29c2-4b29-a02a-2b58036fe875\n",
       "3    01a5dd09-db87-47ac-be78-506c690c4efc\n",
       "4    020d39f9-9375-4377-882e-db83d912aeb7\n",
       "5    02583626-682b-4374-874a-99bd2e6a956e\n",
       "...                                   ...\n",
       "2540 fb29b70b-65af-4bd5-8c78-d33af8cefeb5\n",
       "2541 fb8afe1d-6596-45a6-a6a4-3d83af03f6d1\n",
       "2542 fc65dfd2-cafa-486e-a7f2-753c20d705a0\n",
       "2543 fdb8ed17-e2f0-460a-bb25-9781d63eabf6\n",
       "2544 fe0d170e-af6e-4420-827b-27b125fec214\n",
       "     analysis_protocol.protocol_core.protocol_id\n",
       "                                     <character>\n",
       "1                               smartseq2_v2.3.0\n",
       "2                               smartseq2_v2.4.0\n",
       "3                               smartseq2_v2.3.0\n",
       "4                               smartseq2_v2.3.0\n",
       "5                               smartseq2_v2.3.0\n",
       "...                                          ...\n",
       "2540                            smartseq2_v2.3.0\n",
       "2541                            smartseq2_v2.3.0\n",
       "2542                            smartseq2_v2.3.0\n",
       "2543                            smartseq2_v2.3.0\n",
       "2544                            smartseq2_v2.3.0\n",
       "     analysis_protocol.provenance.document_id\n",
       "                                  <character>\n",
       "1        38f726ad-86fc-404a-97fc-2ac16e6d8461\n",
       "2        1dce56e1-7ae5-4ac4-8002-ebf9c9f8b94d\n",
       "3        0e47010d-03b2-48b7-81a6-271bd3ba09d8\n",
       "4        533acec2-bd51-4dc6-976d-dfbc1a6d725a\n",
       "5        4d8b91cb-655f-41e9-89f9-71051ef84ca6\n",
       "...                                       ...\n",
       "2540     ff64ee51-a381-4646-98ee-596d8fc9163b\n",
       "2541     bcbdc31d-4211-4293-b724-95fa6f92a152\n",
       "2542     917f74ea-be48-4a2d-a7cd-eed1a782e836\n",
       "2543     f76bda5e-4a2a-48f5-8fc8-91cf404c8321\n",
       "2544     6a1bc74e-6cb5-4516-95de-3c380759a39c\n",
       "     analysis_working_group_approval_status\n",
       "                                <character>\n",
       "1                                   blessed\n",
       "2                                   blessed\n",
       "3                                   blessed\n",
       "4                                   blessed\n",
       "5                                   blessed\n",
       "...                                     ...\n",
       "2540                                blessed\n",
       "2541                                blessed\n",
       "2542                                blessed\n",
       "2543                                blessed\n",
       "2544                                blessed\n",
       "     cell_suspension.provenance.document_id derived_organ_label\n",
       "                                <character>         <character>\n",
       "1      00ca0d37-b787-41a4-be59-2aff5b13b0bd            pancreas\n",
       "2      0103aed0-29c2-4b29-a02a-2b58036fe875            pancreas\n",
       "3      01a5dd09-db87-47ac-be78-506c690c4efc            pancreas\n",
       "4      020d39f9-9375-4377-882e-db83d912aeb7            pancreas\n",
       "5      02583626-682b-4374-874a-99bd2e6a956e            pancreas\n",
       "...                                     ...                 ...\n",
       "2540   fb29b70b-65af-4bd5-8c78-d33af8cefeb5            pancreas\n",
       "2541   fb8afe1d-6596-45a6-a6a4-3d83af03f6d1            pancreas\n",
       "2542   fc65dfd2-cafa-486e-a7f2-753c20d705a0            pancreas\n",
       "2543   fdb8ed17-e2f0-460a-bb25-9781d63eabf6            pancreas\n",
       "2544   fe0d170e-af6e-4420-827b-27b125fec214            pancreas\n",
       "     derived_organ_ontology derived_organ_part_label\n",
       "                <character>              <character>\n",
       "1            UBERON:0001264                      nan\n",
       "2            UBERON:0001264                      nan\n",
       "3            UBERON:0001264                      nan\n",
       "4            UBERON:0001264                      nan\n",
       "5            UBERON:0001264                      nan\n",
       "...                     ...                      ...\n",
       "2540         UBERON:0001264                      nan\n",
       "2541         UBERON:0001264                      nan\n",
       "2542         UBERON:0001264                      nan\n",
       "2543         UBERON:0001264                      nan\n",
       "2544         UBERON:0001264                      nan\n",
       "     derived_organ_part_ontology donor_organism.development_stage.ontology\n",
       "                     <character>                               <character>\n",
       "1                            nan                            HSAPDV:0000087\n",
       "2                            nan                            HSAPDV_0000174\n",
       "3                            nan                            HSAPDV_0000099\n",
       "4                            nan                            HSAPDV:0000088\n",
       "5                            nan                            HSAPDV:0000090\n",
       "...                          ...                                       ...\n",
       "2540                         nan                            HSAPDV:0000088\n",
       "2541                         nan                            HSAPDV:0000087\n",
       "2542                         nan                            HSAPDV:0000091\n",
       "2543                         nan                            HSAPDV:0000090\n",
       "2544                         nan                            HSAPDV:0000090\n",
       "     donor_organism.development_stage.ontology_label\n",
       "                                         <character>\n",
       "1                                                nan\n",
       "2                                                nan\n",
       "3                                                nan\n",
       "4                                                nan\n",
       "5                                                nan\n",
       "...                                              ...\n",
       "2540                                             nan\n",
       "2541                                             nan\n",
       "2542                                             nan\n",
       "2543                                             nan\n",
       "2544                                             nan\n",
       "     donor_organism.diseases.ontology donor_organism.diseases.ontology_label\n",
       "                          <character>                            <character>\n",
       "1                        PATO:0000461                                 normal\n",
       "2                        PATO:0000461                                 normal\n",
       "3                        PATO:0000461                                 normal\n",
       "4                        PATO:0000461                                 normal\n",
       "5                        PATO:0000461                                 normal\n",
       "...                               ...                                    ...\n",
       "2540                     PATO:0000461                                 normal\n",
       "2541                     PATO:0000461                                 normal\n",
       "2542                     PATO:0000461                                 normal\n",
       "2543                     PATO:0000461                                 normal\n",
       "2544                     PATO:0000461                                 normal\n",
       "     donor_organism.human_specific.ethnicity.ontology\n",
       "                                          <character>\n",
       "1                                      HANCESTRO:0005\n",
       "2                                      HANCESTRO:0016\n",
       "3                                      HANCESTRO:0005\n",
       "4                                      HANCESTRO:0008\n",
       "5                                      HANCESTRO:0016\n",
       "...                                               ...\n",
       "2540                                   HANCESTRO:0008\n",
       "2541                                   HANCESTRO:0005\n",
       "2542                                   HANCESTRO:0005\n",
       "2543                                   HANCESTRO:0016\n",
       "2544                                   HANCESTRO:0016\n",
       "     donor_organism.human_specific.ethnicity.ontology_label\n",
       "                                                <character>\n",
       "1                                                  European\n",
       "2                        African American or Afro-Caribbean\n",
       "3                                                  European\n",
       "4                                                     Asian\n",
       "5                        African American or Afro-Caribbean\n",
       "...                                                     ...\n",
       "2540                                                  Asian\n",
       "2541                                               European\n",
       "2542                                               European\n",
       "2543                     African American or Afro-Caribbean\n",
       "2544                     African American or Afro-Caribbean\n",
       "                                                    dss_bundle_fqid\n",
       "                                                        <character>\n",
       "1    1f578cdc-144a-44f1-936c-52fbbc6f71b8.2019-05-14T124411.708000Z\n",
       "2    018a3756-0d49-4399-9e56-bc55375ab618.2019-05-30T211857.243000Z\n",
       "3    463a2bc6-e538-453a-bc22-d158c1ed8fb7.2019-05-14T122714.497000Z\n",
       "4    192c10f3-f3a9-464a-a3fe-ef14f7e43e70.2019-05-14T121850.763000Z\n",
       "5    004ad9a7-240b-4521-96c7-c78a8fec769c.2019-05-14T113159.892000Z\n",
       "...                                                             ...\n",
       "2540 8a83810c-22f1-4054-8703-4fc93b6e1fad.2019-05-14T122306.180000Z\n",
       "2541 12ad2c8a-cd50-43c1-b435-fe8d6d7a7d1b.2019-05-14T101534.565000Z\n",
       "2542 1b751146-7141-41bf-a190-bc89561305b2.2019-05-14T122012.383000Z\n",
       "2543 66fe08f2-e937-4632-ae82-77067a10aecf.2019-05-14T120442.136000Z\n",
       "2544 0173919f-8e3d-4982-b6e1-6acbe9e472b2.2019-05-14T115905.298000Z\n",
       "     genes_detected library_preparation_protocol.end_bias\n",
       "          <integer>                           <character>\n",
       "1              6924                           full length\n",
       "2              3171                           full length\n",
       "3              3838                           full length\n",
       "4              4111                           full length\n",
       "5              5834                           full length\n",
       "...             ...                                   ...\n",
       "2540           4802                           full length\n",
       "2541           6215                           full length\n",
       "2542           5992                           full length\n",
       "2543           5017                           full length\n",
       "2544           7188                           full length\n",
       "     library_preparation_protocol.input_nucleic_acid_molecule.ontology\n",
       "                                                           <character>\n",
       "1                                                          OBI:0000869\n",
       "2                                                          OBI:0000869\n",
       "3                                                          OBI:0000869\n",
       "4                                                          OBI:0000869\n",
       "5                                                          OBI:0000869\n",
       "...                                                                ...\n",
       "2540                                                       OBI:0000869\n",
       "2541                                                       OBI:0000869\n",
       "2542                                                       OBI:0000869\n",
       "2543                                                       OBI:0000869\n",
       "2544                                                       OBI:0000869\n",
       "     library_preparation_protocol.input_nucleic_acid_molecule.ontology_label\n",
       "                                                                 <character>\n",
       "1                                                          polyA RNA extract\n",
       "2                                                          polyA RNA extract\n",
       "3                                                          polyA RNA extract\n",
       "4                                                          polyA RNA extract\n",
       "5                                                          polyA RNA extract\n",
       "...                                                                      ...\n",
       "2540                                                       polyA RNA extract\n",
       "2541                                                       polyA RNA extract\n",
       "2542                                                       polyA RNA extract\n",
       "2543                                                       polyA RNA extract\n",
       "2544                                                       polyA RNA extract\n",
       "     library_preparation_protocol.library_construction_method.ontology\n",
       "                                                           <character>\n",
       "1                                                          EFO:0008931\n",
       "2                                                          EFO:0008931\n",
       "3                                                          EFO:0008931\n",
       "4                                                          EFO:0008931\n",
       "5                                                          EFO:0008931\n",
       "...                                                                ...\n",
       "2540                                                       EFO:0008931\n",
       "2541                                                       EFO:0008931\n",
       "2542                                                       EFO:0008931\n",
       "2543                                                       EFO:0008931\n",
       "2544                                                       EFO:0008931\n",
       "     library_preparation_protocol.library_construction_method.ontology_label\n",
       "                                                                 <character>\n",
       "1                                                                 Smart-seq2\n",
       "2                                                                 Smart-seq2\n",
       "3                                                                 Smart-seq2\n",
       "4                                                                 Smart-seq2\n",
       "5                                                                 Smart-seq2\n",
       "...                                                                      ...\n",
       "2540                                                              Smart-seq2\n",
       "2541                                                              Smart-seq2\n",
       "2542                                                              Smart-seq2\n",
       "2543                                                              Smart-seq2\n",
       "2544                                                              Smart-seq2\n",
       "     library_preparation_protocol.provenance.document_id\n",
       "                                             <character>\n",
       "1                   3ab6b486-f900-4f70-ab34-98859ac5f77a\n",
       "2                   3ab6b486-f900-4f70-ab34-98859ac5f77a\n",
       "3                   3ab6b486-f900-4f70-ab34-98859ac5f77a\n",
       "4                   3ab6b486-f900-4f70-ab34-98859ac5f77a\n",
       "5                   3ab6b486-f900-4f70-ab34-98859ac5f77a\n",
       "...                                                  ...\n",
       "2540                3ab6b486-f900-4f70-ab34-98859ac5f77a\n",
       "2541                3ab6b486-f900-4f70-ab34-98859ac5f77a\n",
       "2542                3ab6b486-f900-4f70-ab34-98859ac5f77a\n",
       "2543                3ab6b486-f900-4f70-ab34-98859ac5f77a\n",
       "2544                3ab6b486-f900-4f70-ab34-98859ac5f77a\n",
       "     library_preparation_protocol.strand\n",
       "                             <character>\n",
       "1                             unstranded\n",
       "2                             unstranded\n",
       "3                             unstranded\n",
       "4                             unstranded\n",
       "5                             unstranded\n",
       "...                                  ...\n",
       "2540                          unstranded\n",
       "2541                          unstranded\n",
       "2542                          unstranded\n",
       "2543                          unstranded\n",
       "2544                          unstranded\n",
       "                  project.project_core.project_short_name\n",
       "                                              <character>\n",
       "1    Single cell transcriptome analysis of human pancreas\n",
       "2    Single cell transcriptome analysis of human pancreas\n",
       "3    Single cell transcriptome analysis of human pancreas\n",
       "4    Single cell transcriptome analysis of human pancreas\n",
       "5    Single cell transcriptome analysis of human pancreas\n",
       "...                                                   ...\n",
       "2540 Single cell transcriptome analysis of human pancreas\n",
       "2541 Single cell transcriptome analysis of human pancreas\n",
       "2542 Single cell transcriptome analysis of human pancreas\n",
       "2543 Single cell transcriptome analysis of human pancreas\n",
       "2544 Single cell transcriptome analysis of human pancreas\n",
       "                                                                                                  project.project_core.project_title\n",
       "                                                                                                                         <character>\n",
       "1    Single cell transcriptome analysis of human pancreas reveals transcriptional signatures of aging and somatic mutation patterns.\n",
       "2    Single cell transcriptome analysis of human pancreas reveals transcriptional signatures of aging and somatic mutation patterns.\n",
       "3    Single cell transcriptome analysis of human pancreas reveals transcriptional signatures of aging and somatic mutation patterns.\n",
       "4    Single cell transcriptome analysis of human pancreas reveals transcriptional signatures of aging and somatic mutation patterns.\n",
       "5    Single cell transcriptome analysis of human pancreas reveals transcriptional signatures of aging and somatic mutation patterns.\n",
       "...                                                                                                                              ...\n",
       "2540 Single cell transcriptome analysis of human pancreas reveals transcriptional signatures of aging and somatic mutation patterns.\n",
       "2541 Single cell transcriptome analysis of human pancreas reveals transcriptional signatures of aging and somatic mutation patterns.\n",
       "2542 Single cell transcriptome analysis of human pancreas reveals transcriptional signatures of aging and somatic mutation patterns.\n",
       "2543 Single cell transcriptome analysis of human pancreas reveals transcriptional signatures of aging and somatic mutation patterns.\n",
       "2544 Single cell transcriptome analysis of human pancreas reveals transcriptional signatures of aging and somatic mutation patterns.\n",
       "           project.provenance.document_id\n",
       "                              <character>\n",
       "1    cddab57b-6868-4be4-806f-395ed9dd635a\n",
       "2    cddab57b-6868-4be4-806f-395ed9dd635a\n",
       "3    cddab57b-6868-4be4-806f-395ed9dd635a\n",
       "4    cddab57b-6868-4be4-806f-395ed9dd635a\n",
       "5    cddab57b-6868-4be4-806f-395ed9dd635a\n",
       "...                                   ...\n",
       "2540 cddab57b-6868-4be4-806f-395ed9dd635a\n",
       "2541 cddab57b-6868-4be4-806f-395ed9dd635a\n",
       "2542 cddab57b-6868-4be4-806f-395ed9dd635a\n",
       "2543 cddab57b-6868-4be4-806f-395ed9dd635a\n",
       "2544 cddab57b-6868-4be4-806f-395ed9dd635a\n",
       "     specimen_from_organism.genus_species.ontology\n",
       "                                       <character>\n",
       "1                                   NCBITAXON:9606\n",
       "2                                   NCBITAXON:9606\n",
       "3                                   NCBITAXON:9606\n",
       "4                                   NCBITAXON:9606\n",
       "5                                   NCBITAXON:9606\n",
       "...                                            ...\n",
       "2540                                NCBITAXON:9606\n",
       "2541                                NCBITAXON:9606\n",
       "2542                                NCBITAXON:9606\n",
       "2543                                NCBITAXON:9606\n",
       "2544                                NCBITAXON:9606\n",
       "     specimen_from_organism.genus_species.ontology_label\n",
       "                                             <character>\n",
       "1                                           Homo sapiens\n",
       "2                                           Homo sapiens\n",
       "3                                           Homo sapiens\n",
       "4                                           Homo sapiens\n",
       "5                                           Homo sapiens\n",
       "...                                                  ...\n",
       "2540                                        Homo sapiens\n",
       "2541                                        Homo sapiens\n",
       "2542                                        Homo sapiens\n",
       "2543                                        Homo sapiens\n",
       "2544                                        Homo sapiens\n",
       "     specimen_from_organism.provenance.document_id\n",
       "                                       <character>\n",
       "1             9c1445a1-7287-410e-bb8a-977a8b8e9b05\n",
       "2             14875995-58ca-42cd-9d37-79f5a1f35270\n",
       "3             56b6cd1e-7c2c-43b0-8124-de4a467550fe\n",
       "4             a1b35ebb-b79e-498f-bfc6-f5b4af5bc719\n",
       "5             1f43dc7a-3f89-42a3-8ed7-ee295e59ccb9\n",
       "...                                            ...\n",
       "2540          a1b35ebb-b79e-498f-bfc6-f5b4af5bc719\n",
       "2541          9c1445a1-7287-410e-bb8a-977a8b8e9b05\n",
       "2542          79926ae9-f27d-4ebe-9646-a0188fc145af\n",
       "2543          1f43dc7a-3f89-42a3-8ed7-ee295e59ccb9\n",
       "2544          1f43dc7a-3f89-42a3-8ed7-ee295e59ccb9"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colData(scle)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For more examples on usage, see the Bioconductor documentation for the following parent classes of `SingleCellLoomExperiment`:\n",
    "\n",
    "- [SummarizedExperiment](http://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html#anatomy-of-a-summarizedexperiment)\n",
    "- [SingleCellExperiment](https://bioconductor.org/packages/devel/bioc/vignettes/scater/inst/doc/vignette-intro.html#3_calculating_a_variety_of_expression_values)\n",
    "- [LoomExperiment](http://bioconductor.org/packages/release/bioc/vignettes/LoomExperiment/inst/doc/LoomExperiment.html)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
